### PR TITLE
Push multi-arch images

### PR DESCRIPTION
**Pull request by ddelange**
#### Motivation

Create multi-platform images for `kserve/rest-proxy` images.

#### Modifications

- build runtime images for platforms `linux/amd64`, `linux/arm64`, `linux/ppc64le`, `linux/s390x`
- utilize Docker build cache for `pip` and `dnf` installs as well as `go build`s
- build on `pull_request` as well as on `push` to catch build breaks during PR review;
   only push images to DockerHub on PR merge (push to `main`)
- use `ubi8/go-toolset:1.18` for developer image instead of `ubi8/ubi-minimal:8.4`
- change runtime image base from `scratch` to `ubi-micro:8.7`
- update deprecated `checkout` action
- add `develop` and `run` targets to `Makefile`
- prevent infinite docker inside docker bug when running `make run fmt` inside the developer container

#### Related PRs

- [x] [`modelmesh`](https://github.com/kserve/modelmesh/) (https://github.com/kserve/modelmesh/pull/80)
- [x] [`modelmesh-runtime-adapter`](https://github.com/kserve/modelmesh-runtime-adapter/) (https://github.com/kserve/modelmesh-runtime-adapter/pull/38)
- [x] [`modelmesh-serving`](https://github.com/kserve/modelmesh-serving/) (https://github.com/kserve/modelmesh-serving/pull/314)
- [x] [`modelmesh-minio-examples`](https://github.com/kserve/modelmesh-minio-examples/) (https://github.com/kserve/modelmesh-minio-examples/pull/3)

Resolves kserve/modelmesh-serving#162

### CODE REVIEW

File: .github/workflows/build-and-push.yml [hunk 1]
```diff
@@ -23,24 +23,50 @@ jobs:
...
-          docker tag ${{ env.IMAGE_NAME }}:latest ${{ env.IMAGE_NAME }}:$VERSION
-          docker push ${{ env.IMAGE_NAME }}:$VERSION
+          echo "VERSION=$VERSION" >> $GITHUB_ENV
+          echo "IMAGE_ID=$IMAGE_ID" >> $GITHUB_ENV
+
+      - name: Build and push
+        uses: docker/build-push-action@v3
+        with:
+          platforms: linux/amd64,linux/arm64/v8,linux/ppc64le,linux/s390x
+          push: ${{ github.event_name == 'push' }}
```
Comment (ddelange) [inline .github/workflows/build-and-push.yml:70]: don't push on `pull_request`

File: .github/workflows/build-and-push.yml [hunk 2]
```diff
@@ -23,24 +23,50 @@ jobs:
   build:
     needs: test
     runs-on: ubuntu-latest
+    if: github.event_name == 'push'
```
Comment (ddelange) [inline .github/workflows/build-and-push.yml:26]: got it, build but don't push
```suggestion
```

**Comment by ckadner** [comment]
Hi @ddelange -- I think we need to float the arch platform into a Docker build variable so that the Dockerfile does not have the plaftform id hardcoded for any tar file we download and install during the Dockerbuild.

https://github.com/kserve/rest-proxy/actions/runs/3930863097/jobs/6721448487#step:7:1580
```
#24 [linux/arm64 develop  3/10] RUN set -eux;     wget -qO go.tgz "https://golang.org/dl/go1.18.9.linux-amd64.tar.gz";     sha256sum *go.tgz;     tar -C /usr/local -xzf go.tgz;     go version
#24 47.87 + go version
#24 47.88 /bin/sh: /usr/local/go/bin/go: No such file or directory
#24 ERROR: process "/bin/sh -c set -eux;     wget -qO go.tgz \"[https://golang.org/dl/go${GOLANG_VERSION}.linux-amd64.tar.gz\](https://golang.org/dl/go$%7BGOLANG_VERSION%7D.linux-amd64.tar.gz/)";     sha256sum *go.tgz;     tar -C /usr/local -xzf go.tgz;     go version" did not complete successfully: exit code: 127

#23 [linux/amd64 build 2/2] RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o /go/bin/server ./proxy/
#23 CANCELED
```

File: Dockerfile [hunk 3]
```diff
@@ -13,85 +13,121 @@
...
+    --mount=type=cache,target=/go/pkg \
+    CGO_ENABLED=0 GO111MODULE=on go build -a -o /go/bin/server ./proxy/
+
 
 ###############################################################################
-# Stage 3: Copy binary to create the smallest final runtime image
+# Stage 3: Copy binaries only to create the smallest final runtime image
 ###############################################################################
-FROM scratch AS runtime
+FROM registry.access.redhat.com/ubi8/ubi-micro:8.7 as runtime
```
Comment (tjohnson31415) [inline Dockerfile:125]: What is the reason for this change?

File: Dockerfile [hunk 4]
```diff
@@ -13,85 +13,121 @@
 # limitations under the License.
 
 ###############################################################################
-# Stage 1: Create the develop, test, and build environment
+# Stage 1: Create the developer image for the BUILDPLATFORM only
 ###############################################################################
-FROM  registry.access.redhat.com/ubi8/ubi-minimal:8.7 AS develop
-
 ARG GOLANG_VERSION=1.18.9
```
[the remark this answers is not in the document]
Comment (ckadner) [reply Dockerfile:18]: Good point

**Comment by ckadner** [comment]
@tjohnson31415 --  the issue had to do with Golang version 1.18 vs 1.17 (1.17 is still used for the runtime adapter):
> `go get` no longer builds or installs packages in module-aware mode. `go get` is now dedicated to adjusting dependencies in `go.mod`. Effectively, the `-d` flag is always enabled. To install the latest version of an executable outside the context of the current module, use `go install example.com/cmd@latest`. Any version query may be used instead of latest. This form of go install was added in Go `1.16`, so projects supporting older versions may need to provide install instructions for both go install and go get. go get now reports an error when used outside a module, since there is no `go.mod` file to update. In `GOPATH` mode (with `GO111MODULE=off`), `go get` still builds and installs packages, as before.

https://tip.golang.org/doc/go1.18#go-get

This meant that the `protoc-gen-go` CLI was actually not installed since the Dockerfile was updated from Go 1.17 to Go 1.18

File: Dockerfile [hunk 5]
```diff
@@ -13,85 +13,127 @@
...
-    nodejs && \
+RUN --mount=type=cache,target=/root/.cache/dnf:rw \
+    dnf install --setopt=cachedir=/root/.cache/dnf -y --nodocs \
+       python3 \
+       python3-pip \
+       nodejs \
+    && true
+
+# Install pre-commit
+ARG PIP_CACHE_DIR=/root/.cache/pip
```
Comment (tjohnson31415) [inline Dockerfile:35]: NIT: move setting this env var into the `RUN` command below since it is not a configurable ARG

File: Dockerfile [hunk 6]
```diff
@@ -13,85 +13,127 @@
...
 # Install go protoc plugins
-ENV PATH /root/go/bin:$PATH
-RUN go get google.golang.org/protobuf/cmd/protoc-gen-go \
-           google.golang.org/grpc/cmd/protoc-gen-go-grpc
-
-WORKDIR /opt/app
+ENV PATH $HOME/go/bin:$PATH
+RUN true \
+    && go get google.golang.org/protobuf/cmd/protoc-gen-go \
+              google.golang.org/grpc/cmd/protoc-gen-go-grpc \
```
[the remark this answers is not in the document]
Comment (ckadner) [reply Dockerfile:83]: without `go get` we end up with a slightly different version:

```
[root@ca4b827f4264 app]# protoc-gen-go --version

protoc-gen-go v1.30.0-devel
```

with the `go get` we get:

```
[root@6c792c90e689 app]# protoc-gen-go --version

protoc-gen-go v1.30.0
```

File: Dockerfile (hunk 5 above)
Comment (ckadner) [reply Dockerfile]: Right, not a build `ARG`. I converted it into `ENV` since it changes a global Python setting. So it should be an environment variable that remains for any future pip installs, not just the one line, even though there are no further pip installs for now.

File: Dockerfile [hunk 7]
```diff
@@ -13,85 +13,127 @@
...
-# Build the binary
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o /go/bin/server ./proxy/
+# https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope
+# don't provide "default" values (e.g. 'ARG TARGETARCH=amd64') for non-buildx environments,
+# see https://github.com/docker/buildx/issues/510
+ARG TARGETOS
+ARG TARGETARCH
+
+ARG GOOS=${TARGETOS:-linux}
+ARG GOARCH=${TARGETARCH:-amd64}
```
[the remark this answers is not in the document]
Comment (ckadner) [reply Dockerfile]: Right. I moved it.

File: Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (tjohnson31415) [reply Dockerfile:83]: Hmm, that is interesting. Thanks for the details. We'll roll with both commands then.

**Review by tjohnson31415 — APPROVED** [review]
LGTM
